src/export-plan/pdf-gen.ts

```diff
@@ -0,0 +1,328 @@
+            if (data.column.index > 0)
+              doc.line(data.cell.x, data.cell.y + data.row.height, data.cell.x, data.cell.y);
+            // draw top horizontal border for last row
+            doc.line(data.cell.x + data.column.width, data.cell.y, data.cell.x, data.cell.y);
+          }
+
+          tableHeight = data.cell.y + data.row.height - startct + rowHeight;
+
+          if (
+            data.column.index === 2 &&
+            data.section === 'body' &&
+            data.row.index >= 0 &&
+            data.row.index < body.length
+          ) {
+            let yPos = data.cell.y + 3;
+            groups[data.row.index].forEach((group, index) => {
+              const xPos =
+                data.cell.x + doc.getTextWidth(body[data.row.index][2].split('\n')[index]) + 8;
+
+              const colour = colours[data.row.index][index];
+              renderBubbles(doc, xPos, yPos, group, colour);
+              yPos += rowFontSize + 1.5;
+            });
+          }
+        },
+      });
+    }
+
+    doc.setDrawColor(216, 216, 216);
+    doc.setLineWidth(0.5);
+    doc.roundedRect(tableX, startct - rowHeight, tableWidth, tableHeight, 4, 4);
+    startct += tableHeight + tableGap;
+  }
+
+  const footerY = doc.internal.pageSize.height - 25;
+  const footerX =
+    (doc.internal.pageSize.width - doc.getTextWidth('downloaded from courseplan.io')) / 2;
+  doc.text('downloaded from ', footerX, footerY, { align: 'left' });
+  doc.textWithLink('courseplan.io', footerX + doc.getTextWidth('downloaded from '), footerY, {
+    url: 'https://courseplan.io',
+  });
+  doc.setDrawColor(0);
+  doc.line(
+    footerX + doc.getTextWidth('downloaded from '),
+    footerY + 3,
+    footerX + doc.getTextWidth('downloaded from courseplan.io'),
+    footerY + 3
+  );
+  const pdfName = `${store.state.userName.firstName}'s CoursePlan`;
+  doc.save(pdfName);
+};
+
+const getCourseRows = (sem: FirestoreSemester): [string[][], string[][], string[][]] => {
+  const filteredCourses = sem.courses.filter(
+    (course): course is FirestoreSemesterCourse => !isPlaceholderCourse(course)
+  );
+  const rows = filteredCourses.map(course => {
+    const [reqs, groups, colours] = getFulfilledReqs(course);
+    return [
+      [`${course.code}: ${course.name}`, course.credits.toString(), reqs.join('\n')],
+      groups,
+      colours,
+    ];
+  });
+
+  return [rows.map(row => row[0]), rows.map(row => row[1]), rows.map(row => row[2])];
+};
+
+const trimEmptySems = (sems: readonly FirestoreSemester[]): readonly FirestoreSemester[] => {
+  let maxNonemptyIndex = -1;
+  for (let i = 0; i < sems.length; i += 1) {
+    if (sems[i].courses.length > 0) maxNonemptyIndex = i;
+  }
+  return sems.slice(0, maxNonemptyIndex + 1);
+};
+
+const getFulfilledReqs = (
+  course: FirestoreSemesterCourse
+): readonly [string[], string[], string[]] => {
+  let reqsFulfilled = store.state.safeRequirementFulfillmentGraph.getConnectedRequirementsFromCourse(
```

Comment:
I think `const` is always preferred over `let`, if possible. So in this example, it would be better to just use a new variable instead of reassigning the variable a mutated value.